pkg/tenant/matcher/matcher.go

```diff
@@ -0,0 +1,121 @@
+package tenantmatcher
```

Comment:
I think we should merge everything inside `pkg/tenant` into `pkg/providers`

Reply:
Furthermore, I would suggest everything inside `pkg/providerloader` should be merged into a subpackage of providers: `pkg/providers/loader`